include/envoy/thread_local/thread_local.h

```diff
@@ -89,6 +89,81 @@ class Slot {
 
 using SlotPtr = std::unique_ptr<Slot>;
 
+// Provides a typesafe API for slots.
+//
+// TODO(jmarantz): Rename the Slot class to something like RawSlot, where the
+// only reference is from TypedSlot, which we can then rename to Slot.
+template <class T> class TypedSlot {
+public:
+  /**
+   * @return true if the TypedSlot object has been allocated a slot.
+   */
+  bool hasSlot() const { return slot_ != nullptr; }
+
+  /**
+   * @param slot The allocated slot object. Ownership transferred into the TypedSlot.
+   */
+  void setSlot(SlotPtr&& slot) { slot_ = std::move(slot); }
```

Comment:
Do we need an explicit setter for this? Why can't this be done at construction time?

Reply:
+1, my advice would actually be to have `allocateSlot()` be templated and return one of these things. Then I think you can avoid set/has?

Reply:
Virtual functions can't be templated, so `allocateSlot` would have to be non-virtual and delegate to a type-unsafe virtual function.

Reply:
Yes, that's what I meant. We do this in various other places.

Reply:
WDYT of having the slot allocator be a required ctor param for TypedSlot?

Not wedded to that, I can check to see if that's compatible at various call-sites.

Reply:
I think you could do it a bunch of different ways. I don't have a strong preference as long as we remove set/has.

Reply:
done